Lesson 2. Create models

AntonBuyanov opened the pull request:
Create models: answer, category, question, test, user. Migrations to limit integrity. Default values in tables: tests, answers

Commit:
feat: Rails models

Create models: answer, category, question, test, user. Migrations to limit integrity. Default values in tables: tests, answers

VidgarVii commented on db/migrate/20220526063154_add_tests_title_null_constraint.rb:
```diff
@@ -0,0 +1,12 @@
+  # end
+
+  def up
+    change_column_null(:tests, :title, false)
```
е стоит добавлять отдельные миграции для изменения полей в одном ПР. Миграции всегда являются опасным местом в ПО. Так как меняют схему базы. И всегда стараются минимизировать кол-во миграций.
В этом ДЗ надо создать 5 моделей и 5 таблиц. Значит у тебя должно быть 5 миграций. Прописывай все нужные опции полям сразу. Все остальные миграции являются обычным оверхедом

Пример
```
t.string :title, null: false 
```
Большое кол-во миграций увеличивает время развёртывания ПО. Так же миграция является местом риска, так как может сломать существующую базу. Еще они имеют свойство протухать и быть не актуальными или не рабочими. По этому всегда стараются минимизировать кол-во миграций.

VidgarVii commented on db/migrate/20220526063154_add_tests_title_null_constraint.rb:
```diff
@@ -0,0 +1,12 @@
+class AddTestsTitleNullConstraint < ActiveRecord::Migration[6.1]
+  # def change
+  # end
```
это мусорный комент

VidgarVii commented on db/schema.rb:
```diff
@@ -0,0 +1,48 @@
+    t.string "body", null: false
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+    t.string "correct"
```
оле `correct` может по сути иметь только 2 значния "да" или "нет"
когда поле подразумевает только 2 значения пользуются типом `boolean` (логическое поле, которое может содержать только `true`/`false` значения)

VidgarVii commented on db/schema.rb:
```diff
@@ -0,0 +1,48 @@
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+  end
+
```
на этом задании надо еще добавить связи м/у таблицами. 
Так как тест принадлежит определенной категории надо использовать метод references
пример

```
t.references :category, null: false, foreign_key: true
```

тут мы укажем, что поле должно быть обязательно заполнен и обеспечим ссылочную целостность данных, через внешний ключ
[внешние ключи](https://en.wikipedia.org/wiki/Foreign_key)


Тест принадлежит Категории и Автору
Тест имеет Вопросы
Вопросы имеют Ответы

Вспомни ДЗ про SQL там уже добавляли эти связи

VidgarVii commented on db/schema.rb:
```diff
@@ -0,0 +1,48 @@
+
+  create_table "tests", force: :cascade do |t|
+    t.string "title", null: false
+    t.integer "level"
```
С помощью механизма миграций добавьте значения по умолчанию:
Для уровня Теста

VidgarVii replied:
null: false

VidgarVii commented on db/schema.rb:
```diff
@@ -0,0 +1,48 @@
+
+  create_table "users", force: :cascade do |t|
+    t.string "name", null: false
+    t.integer "number"
```
какие данные будет содержать поле number?

Commit:
fix: Rails models

Removed unnecessary migrations. In the response table, the attribute correct is boolean. Links between tables. Default for the tests level.

VidgarVii commented on db/migrate/20220527085210_create_tests.rb:
```diff
@@ -0,0 +1,12 @@
+      t.string :title, null: false
+      t.integer :level, null: false, :default => 1
+      t.references :category, null: false, foreign_key: true
+      t.references :author, null: false, foreign_key: true
```
Посмотри какой ключ был создан. Он будет ссылаться на таблицу авторы. У нас такой таблицы нет. Надо чтобы внешний ключ смотрел в таблицу бзеры. Но название проекта должно остаться прежним.

VidgarVii commented on db/migrate/20220527085908_create_answers.rb:
```diff
@@ -0,0 +1,11 @@
+  def change
+    create_table :answers do |t|
+      t.string :body, null: false
+      t.boolean :correct
```
Для булевых полей принято добавлять опцию null false. Чтобы оно имело только 2 значения правди или ложь

Commit:
fix: Rails models

The tests table has the users foreign key. Integrity constraint for the correct column in the answer table.
